### PR TITLE
[Storage][DataMovement] Add ProviderId to resources and job plan file

### PR DESCRIPTION
**Pull request by jalauzon-msft**
This change adds the concept of `ProviderId` where each resource type now has a string id that represents which provider, they need for resuming. This information is stored in the job plan file for source and destination and used when choosing the provider to use for resume.

### CODE REVIEW

File: sdk/storage/Azure.Storage.DataMovement/tests/JobPlanHeaderTests.cs [hunk 1]
```diff
@@ -3,6 +3,7 @@
 
 using System.IO;
 using Azure.Storage.DataMovement.JobPlan;
+using FastSerialization;
```
Comment (amnguye) [inline sdk/storage/Azure.Storage.DataMovement/tests/JobPlanHeaderTests.cs:6]: was this added by mistake?

**Review by amnguye — APPROVED** [review]

File: sdk/storage/Azure.Storage.DataMovement/tests/JobPlanHeaderTests.cs (hunk 1 above)
Comment (jalauzon-msft) [reply sdk/storage/Azure.Storage.DataMovement/tests/JobPlanHeaderTests.cs]: Yep. Stupid VS adding imports when you start typing stuff and it thinks it knows what you want.

File: sdk/storage/Azure.Storage.DataMovement/api/Azure.Storage.DataMovement.netstandard2.0.cs [hunk 2]
```diff
@@ -113,6 +115,7 @@ public abstract partial class StorageResource
     {
         protected StorageResource() { }
         protected internal abstract bool IsContainer { get; }
+        public abstract string ProviderId { get; }
```
[the remark this answers is not in the document]
Comment (jalauzon-msft) [reply sdk/storage/Azure.Storage.DataMovement/api/Azure.Storage.DataMovement.netstandard2.0.cs:118]: Discussed offline.

**Pull request merged**